src/crush/CrushWrapper.cc

```diff
@@ -1370,6 +1370,33 @@ int CrushWrapper::rebuild_roots_with_classes()
   return trim_roots_with_class(true);
 }
 
+void CrushWrapper::get_incompat_choose_args(std::map<__u32, crush_choose_arg*> &i2choose_arg)
+{
+  if (HAVE_FEATURE(CRUSH_CHOOSEARGS))
+    return;
+  if (choose_args.size() != 1)
+    return;
+  crush_choose_arg_map arg_map = choose_args.begin()->second;
+  for (__u32 i = 0; i < arg_map.size; i++) {
+    crush_choose_arg *arg = &arg_map.args[i];
+    if (arg->weight_set_size == 0 &&
+	arg->ids_size == 0)
+	continue;
+    if (arg->weight_set_size != 1)
+      return;
+    if (arg->ids_size != 0)
+      return;
+  }
+
+  for (__u32 i = 0; i < arg_map.size; i++) {
+    crush_choose_arg *arg = &arg_map.args[i];
+    if (arg->weight_set_size == 0 &&
+	arg->ids_size == 0)
+	continue;
+    id2choose_arg[i] = arg;
+  }
```

Comment:
it seems like this second part is just repackaging an existing vector/array as a map.  the important part is the compatibility check that happens above.  maybe recast this function to return bool (can_encode_compat_choose_args), do a check vs features (and assert out if it's not compat but the feature is missing).  then in the straw2 encode if a bool should_encode_compat_choose_args is true look up the bucket weights directly in the arg_map[bucketno] array?